mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java

```diff
@@ -549,6 +582,74 @@ public void setFlagsShouldUpdateFlagsWhenMessageIsInTwoMailboxes() throws Except
+        message3.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message3);
+        message4.setUid(mapperProvider.generateMessageUid());
+        message4.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message4);
+
+        MessageId messageId = message1.getMessageId();
+        sut.setFlags(messageId, ImmutableList.of(message1.getMailboxId()), new Flags(Flag.ANSWERED), FlagsUpdateMode.ADD);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+    }
+
+    @ContractTest
+    public void setFlagsShouldWorkWhenCalledOnDuplicatedMailbox() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        message1.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message1);
+        message2.setUid(mapperProvider.generateMessageUid());
+        message2.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message2);
+        message3.setUid(mapperProvider.generateMessageUid());
+        message3.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message3);
+        message4.setUid(mapperProvider.generateMessageUid());
+        message4.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message4);
+
+        MessageId messageId = message1.getMessageId();
+        sut.setFlags(messageId, ImmutableList.of(message1.getMailboxId(), message1.getMailboxId()), new Flags(Flag.ANSWERED), FlagsUpdateMode.ADD);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+    }
+
+    @ContractTest
+    public void setFlagsShouldWorkWithConcurency() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        message1.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message1);
+
+        int threadCount = 2;
+        int updateCount = 10;
+        CountDownLatch countDownLatch = new CountDownLatch(threadCount);
+        ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
+        for (int i = 0; i < threadCount; i++) {
+            executorService.submit(new ConcurentSetFlagTestRunnable(i, updateCount, message1, countDownLatch));
+        }
+        executorService.shutdown();
+        assertThat(executorService.awaitTermination(1, TimeUnit.MINUTES))
+            .isTrue();
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(message1.getMessageId()), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).createFlags().getUserFlags()).hasSize(threadCount * updateCount);
+    }
```

Comment:
Do you think you can add the same tests with some delete operations instead of add?